compute/main.go

```diff
@@ -0,0 +1,656 @@
+	resourceGroupName = "sample-resource-group"
+	vmName            = "sample-vm"
+	vnetName          = "sample-vnet"
+	subnetName        = "sample-subnet"
+	nsgName           = "sample-nsg"
+	nicName           = "sample-nic"
+	diskName          = "sample-disk"
+	publicIPName      = "sample-public-ip"
+)
+
+func main() {
+	subscriptionId = os.Getenv("AZURE_SUBSCRIPTION_ID")
+	if len(subscriptionId) == 0 {
+		log.Println("AZURE_SUBSCRIPTION_ID is not set.")
+		return
+	}
+	log.Println("AZURE_SUBSCRIPTION_ID:", subscriptionId)
+	//create virtual machine
+	createVM()
+
+	//delete virtual machine
+	// deleteVM()
+}
+
+func createVM() {
+
+	conn, err := connectionAzure()
+	handlerErr(err)
+	ctx := context.Background()
+
+	log.Println("start creating virtual machine...")
+	resourceGroup, err := createResourceGroup(ctx, conn, resourceGroupName)
+	handlerErr(err)
+	toString(resourceGroup)
+
+	virtualNetwork, err := createVirtualNetwork(ctx, conn)
+	handlerErr(err)
+	toString(virtualNetwork)
+
+	subnet, err := createSubnets(ctx, conn, vnetName, subnetName)
+	handlerErr(err)
+	toString(subnet)
+
+	publicIP, err := createPublicIP(ctx, conn, publicIPName)
+	handlerErr(err)
+	toString(publicIP)
+
+	// network security group
+	nsg, err := createNetworkSecurityGroup(ctx, conn, nsgName)
+	handlerErr(err)
+	toString(nsg)
+
+	subnetID := subnet.ID
+	publicID := publicIP.ID
+	networkSecurityGroupID := nsg.ID
+	netWorkInterface, err := createNetWorkInterface(ctx, conn, subnetID, publicID, networkSecurityGroupID)
+	handlerErr(err)
+	toString(netWorkInterface)
+
+	networkInterfaceID := netWorkInterface.ID
+	virtualMachine, err := createVirtualMachine(ctx, conn, *networkInterfaceID)
+	handlerErr(err)
+	toString(virtualMachine)
+
+	log.Println("success created virtual machine")
+}
+
+func deleteVM() {
+	conn, err := connectionAzure()
+	handlerErr(err)
+	ctx := context.Background()
+
+	log.Println("start deleting virtual machine...")
+	_, err = deleteVirtualMachine(ctx, conn)
+	handlerErr(err)
+	log.Println("deleted virtual machine")
+
+	_, err = deleteDisk(ctx, conn, diskName)
+	handlerErr(err)
+	log.Println("deleted disk")
+
+	_, err = deleteNetWorkInterface(ctx, conn)
+	handlerErr(err)
+	log.Println("deleted network interface")
+
+	_, err = deleteNetworkSecurityGroup(ctx, conn, nsgName)
+	handlerErr(err)
+	log.Println("deleted network security group")
+
+	_, err = deletePublicIP(ctx, conn, publicIPName)
+	handlerErr(err)
+	log.Println("deleted public IP")
+
+	_, err = deleteSubnets(ctx, conn, vnetName, subnetName)
+	handlerErr(err)
+	log.Println("deleted subnet")
+
+	_, err = deleteVirtualNetWork(ctx, conn)
+	handlerErr(err)
+	log.Println("deleted virtual network")
+
+	_, err = deleteResourceGroup(ctx, conn)
+	handlerErr(err)
+	log.Println("deleted resource group")
+	log.Println("success deleted virtual machine.")
+}
+
+func handlerErr(err error) {
+	if err != nil {
+		panic(err)
+	}
+}
+
+func toString(marshaler json.Marshaler) {
+	data, err := marshaler.MarshalJSON()
```

Comment:
Usually if we want to marshal something to JSON, we just need to use this
```go
b, err := json.Marshal(v)
```
and print it
```go
log.Printf(string(b))
```
This implementation might throw errors when the instance you passed does not implements the `json.Marshaler` interface (this might happen if the model does not have a custom marshaler)